File: mydia-rs/crates/indexers/src/category_mapping.rs
//! Torznab category mapping.
//!
//! Port of `lib/mydia/indexers/category_mapping.ex`. Newznab-style
//! numeric category IDs.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LibraryType {
    Movies,
    Series,
    Music,
    Books,
    Adult,
    Mixed,
    Other,
}

// Movies (2000-2999)
pub const MOVIES_PARENT: i32 = 2000;
pub const MOVIES_FOREIGN: i32 = 2010;
pub const MOVIES_OTHER: i32 = 2020;
pub const MOVIES_SD: i32 = 2030;
pub const MOVIES_HD: i32 = 2040;
pub const MOVIES_UHD: i32 = 2045;
pub const MOVIES_BLURAY: i32 = 2050;
pub const MOVIES_3D: i32 = 2060;
pub const MOVIES_DVD: i32 = 2070;

// Audio/Music (3000-3999)
pub const AUDIO_PARENT: i32 = 3000;
pub const AUDIO_MP3: i32 = 3010;
pub const AUDIO_VIDEO: i32 = 3020;
pub const AUDIO_AUDIOBOOK: i32 = 3030;
pub const AUDIO_LOSSLESS: i32 = 3040;
pub const AUDIO_OTHER: i32 = 3050;
pub const AUDIO_FOREIGN: i32 = 3060;

// TV (5000-5999)
pub const TV_PARENT: i32 = 5000;
pub const TV_WEBDL: i32 = 5010;
pub const TV_FOREIGN: i32 = 5020;
pub const TV_SD: i32 = 5030;
pub const TV_HD: i32 = 5040;
pub const TV_UHD: i32 = 5045;
pub const TV_OTHER: i32 = 5050;
pub const TV_SPORT: i32 = 5060;
pub const TV_ANIME: i32 = 5070;
pub const TV_DOCUMENTARY: i32 = 5080;

// Adult/XXX (6000-6999)
pub const XXX_PARENT: i32 = 6000;
pub const XXX_DVD: i32 = 6010;
pub const XXX_WMV: i32 = 6020;
pub const XXX_XVID: i32 = 6030;
pub const XXX_X264: i32 = 6040;
pub const XXX_UHD: i32 = 6045;
pub const XXX_PACK: i32 = 6050;
pub const XXX_IMAGESET: i32 = 6060;
pub const XXX_PACKS: i32 = 6070;
pub const XXX_SD: i32 = 6080;
pub const XXX_WEBDL: i32 = 6090;

// Books (7000-7999)
pub const BOOKS_PARENT: i32 = 7000;
pub const BOOKS_FOREIGN: i32 = 7010;
pub const BOOKS_EBOOK: i32 = 7020;
pub const BOOKS_COMICS: i32 = 7030;
pub const BOOKS_MAGAZINES: i32 = 7040;
pub const BOOKS_TECHNICAL: i32 = 7050;
pub const BOOKS_OTHER: i32 = 7060;
pub const BOOKS_AUDIOBOOK: i32 = 7070;

// Other (8000-8999)
pub const OTHER_PARENT: i32 = 8000;
pub const OTHER_MISC: i32 = 8010;
pub const OTHER_HASHED: i32 = 8020;

/// All Torznab category IDs for the given [`LibraryType`].
#[must_use]
pub fn categories_for_type(t: LibraryType) -> Vec<i32> {
    match t {
        LibraryType::Movies => vec![
            MOVIES_PARENT,
            MOVIES_FOREIGN,
            MOVIES_OTHER,
            MOVIES_SD,
            MOVIES_HD,
            MOVIES_UHD,
            MOVIES_BLURAY,
            MOVIES_3D,
            MOVIES_DVD,
        ],
        LibraryType::Series => vec![
            TV_PARENT,
            TV_WEBDL,
            TV_FOREIGN,
            TV_SD,
            TV_HD,
            TV_UHD,
            TV_OTHER,
            TV_SPORT,
            TV_ANIME,
            TV_DOCUMENTARY,
        ],
        LibraryType::Music => vec![
            AUDIO_PARENT,
            AUDIO_MP3,
            AUDIO_VIDEO,
            AUDIO_AUDIOBOOK,
            AUDIO_LOSSLESS,
            AUDIO_OTHER,
            AUDIO_FOREIGN,
        ],
        LibraryType::Books => vec![
            BOOKS_PARENT,
            BOOKS_FOREIGN,
            BOOKS_EBOOK,
            BOOKS_COMICS,
            BOOKS_MAGAZINES,
            BOOKS_TECHNICAL,
            BOOKS_OTHER,
            BOOKS_AUDIOBOOK,
        ],
        LibraryType::Adult => vec![
            XXX_PARENT,
            XXX_DVD,
            XXX_WMV,
            XXX_XVID,
            XXX_X264,
            XXX_UHD,
            XXX_PACK,
            XXX_IMAGESET,
            XXX_PACKS,
            XXX_SD,
            XXX_WEBDL,
        ],
        LibraryType::Mixed => {
            let mut out = categories_for_type(LibraryType::Movies);
            out.extend(categories_for_type(LibraryType::Series));
            out
        }
        LibraryType::Other => Vec::new(),
    }
}

/// Parent (non-subcategory) ID for a [`LibraryType`].
#[must_use]
pub fn parent_category(t: LibraryType) -> Option<i32> {
    match t {
        LibraryType::Movies => Some(MOVIES_PARENT),
        LibraryType::Series => Some(TV_PARENT),
        LibraryType::Music => Some(AUDIO_PARENT),
        LibraryType::Books => Some(BOOKS_PARENT),
        LibraryType::Adult => Some(XXX_PARENT),
        LibraryType::Mixed | LibraryType::Other => None,
    }
}

/// Human-readable label for a category ID.
#[must_use]
pub fn category_name(id: i32) -> String {
    match id {
        MOVIES_PARENT => "Movies".into(),
        MOVIES_FOREIGN => "Movies/Foreign".into(),
        MOVIES_OTHER => "Movies/Other".into(),
        MOVIES_SD => "Movies/SD".into(),
        MOVIES_HD => "Movies/HD".into(),
        MOVIES_UHD => "Movies/UHD".into(),
        MOVIES_BLURAY => "Movies/BluRay".into(),
        MOVIES_3D => "Movies/3D".into(),
        MOVIES_DVD => "Movies/DVD".into(),
        AUDIO_PARENT => "Audio".into(),
        AUDIO_MP3 => "Audio/MP3".into(),
        AUDIO_VIDEO => "Audio/Video".into(),
        AUDIO_AUDIOBOOK => "Audio/Audiobook".into(),
        AUDIO_LOSSLESS => "Audio/Lossless".into(),
        AUDIO_OTHER => "Audio/Other".into(),
        AUDIO_FOREIGN => "Audio/Foreign".into(),
        TV_PARENT => "TV".into(),
        TV_WEBDL => "TV/WEB-DL".into(),
        TV_FOREIGN => "TV/Foreign".into(),
        TV_SD => "TV/SD".into(),
        TV_HD => "TV/HD".into(),
        TV_UHD => "TV/UHD".into(),
        TV_OTHER => "TV/Other".into(),
        TV_SPORT => "TV/Sport".into(),
        TV_ANIME => "TV/Anime".into(),
        TV_DOCUMENTARY => "TV/Documentary".into(),
        XXX_PARENT => "XXX".into(),
        XXX_DVD => "XXX/DVD".into(),
        XXX_WMV => "XXX/WMV".into(),
        XXX_XVID => "XXX/XviD".into(),
        XXX_X264 => "XXX/x264".into(),
        XXX_UHD => "XXX/UHD".into(),
        XXX_PACK => "XXX/Pack".into(),
        XXX_IMAGESET => "XXX/ImageSet".into(),
        XXX_PACKS => "XXX/Packs".into(),
        XXX_SD => "XXX/SD".into(),
        XXX_WEBDL => "XXX/WEB-DL".into(),
        BOOKS_PARENT => "Books".into(),
        BOOKS_FOREIGN => "Books/Foreign".into(),
        BOOKS_EBOOK => "Books/EBook".into(),
        BOOKS_COMICS => "Books/Comics".into(),
        BOOKS_MAGAZINES => "Books/Magazines".into(),
        BOOKS_TECHNICAL => "Books/Technical".into(),
        BOOKS_OTHER => "Books/Other".into(),
        BOOKS_AUDIOBOOK => "Books/Audiobook".into(),
        OTHER_PARENT => "Other".into(),
        OTHER_MISC => "Other/Misc".into(),
        OTHER_HASHED => "Other/Hashed".into(),
        other => format!("Unknown ({other})"),
    }
}

#[must_use]
pub fn category_matches_type(id: i32, t: LibraryType) -> bool {
    categories_for_type(t).contains(&id)
}

#[must_use]
pub fn type_for_category(id: i32) -> LibraryType {
    match id {
        i if (2000..3000).contains(&i) => LibraryType::Movies,
        i if (3000..4000).contains(&i) => LibraryType::Music,
        i if (5000..6000).contains(&i) => LibraryType::Series,
        i if (6000..7000).contains(&i) => LibraryType::Adult,
        i if (7000..8000).contains(&i) => LibraryType::Books,
        _ => LibraryType::Other,
    }
}

/// Resolve a Cardigann category name to a Torznab ID. The match is
/// case-insensitive on the lowercased form (`"Movies/HD"` → `2040`).
#[must_use]
#[allow(clippy::too_many_lines)]
pub fn category_id_from_name(name: &str) -> Option<i32> {
    let normalized = name.to_lowercase();
    let id = match normalized.as_str() {
        // Movies
        "movies" => MOVIES_PARENT,
        "movies/foreign" => MOVIES_FOREIGN,
        "movies/other" => MOVIES_OTHER,
        "movies/sd" => MOVIES_SD,
        "movies/hd" => MOVIES_HD,
        "movies/uhd" => MOVIES_UHD,
        "movies/bluray" | "movies/blu-ray" => MOVIES_BLURAY,
        "movies/3d" => MOVIES_3D,
        "movies/dvd" => MOVIES_DVD,
        "movies/divx/xvid" | "movies/mp4" | "movies/svcd/vcd" => MOVIES_SD,
        "movies/h.264/x264" | "movies/hevc/x265" => MOVIES_HD,
        "movies/dubs/dual audio" | "movies/bollywood" => MOVIES_FOREIGN,
        // Audio
        "audio" | "music" | "music/dvd" | "music/album" | "music/box set" | "music/discography"
        | "music/single" => AUDIO_PARENT,
        "audio/mp3" | "music/mp3" | "music/aac" => AUDIO_MP3,
        "audio/video" | "music/video" | "music/concerts" => AUDIO_VIDEO,
        "audio/audiobook" => AUDIO_AUDIOBOOK,
        "audio/lossless" | "music/lossless" => AUDIO_LOSSLESS,
        "audio/other" | "music/radio" | "music/other" => AUDIO_OTHER,
        "audio/foreign" => AUDIO_FOREIGN,
        // TV
        "tv" | "tv/dvd" => TV_PARENT,
        "tv/web-dl" | "tv/webdl" => TV_WEBDL,
        "tv/foreign" => TV_FOREIGN,
        "tv/sd" | "tv/divx/xvid" | "tv/svcd/vcd" => TV_SD,
        "tv/hd" | "tv/hevc/x265" => TV_HD,
        "tv/uhd" => TV_UHD,
        "tv/other" => TV_OTHER,
        "tv/sport" => TV_SPORT,
        "tv/anime" | "tv/cartoons" => TV_ANIME,
        "tv/documentary" => TV_DOCUMENTARY,
        // Adult
        "xxx" | "xxx/magazine" | "xxx/hentai" | "xxx/games" => XXX_PARENT,
        "xxx/dvd" | "xxx/video" => XXX_DVD,
        "xxx/wmv" => XXX_WMV,
        "xxx/xvid" => XXX_XVID,
        "xxx/x264" => XXX_X264,
        "xxx/uhd" => XXX_UHD,
        "xxx/pack" => XXX_PACK,
        "xxx/imageset" | "xxx/picture" => XXX_IMAGESET,
        "xxx/packs" => XXX_PACKS,
        "xxx/sd" => XXX_SD,
        "xxx/web-dl" => XXX_WEBDL,
        // Books
        "books" => BOOKS_PARENT,
        "books/foreign" => BOOKS_FOREIGN,
        "books/ebook" | "other/e-books" => BOOKS_EBOOK,
        "books/comics" | "other/comics" => BOOKS_COMICS,
        "books/magazines" => BOOKS_MAGAZINES,
        "books/technical" | "other/tutorial" => BOOKS_TECHNICAL,
        "books/other" => BOOKS_OTHER,
        "books/audiobook" | "other/audiobook" => BOOKS_AUDIOBOOK,
        // Other / PC / Console
        "other"
        | "other/emulation"
        | "other/images"
        | "other/mobile phone"
        | "other/nulled script"
        | "pc"
        | "pc/games"
        | "pc/mac"
        | "pc/mobile-android"
        | "pc/mobile-ios"
        | "pc/mobile-other"
        | "console"
        | "console/ps3"
        | "console/ps4"
        | "console/psp"
        | "console/xbox"
        | "console/xbox 360"
        | "console/wii"
        | "console/nds"
        | "console/3ds"
        | "console/other" => OTHER_PARENT,
        "other/misc" | "other/other" => OTHER_MISC,
        "other/hashed" => OTHER_HASHED,
        "other/sounds" => AUDIO_OTHER,
        _ => return None,
    };
    Some(id)
}

/// Site-specific category ID → Torznab via a definition's
/// `categorymappings` block.
#[must_use]
pub fn map_site_category_to_torznab(
    site_category_id: Option<&str>,
    mappings: &[serde_json::Value],
) -> Option<i32> {
    let id = site_category_id?;
    let id_int: Option<i64> = id.parse().ok();

    let matched = mappings.iter().find(|m| {
        let mid = m.get("id");
        match mid {
            Some(serde_json::Value::String(s)) => s == id,
            Some(serde_json::Value::Number(n)) => match id_int {
                Some(want) => n.as_i64() == Some(want),
                None => n.to_string() == id,
            },
            _ => false,
        }
    })?;

    let cat = matched.get("cat").and_then(|v| v.as_str())?;
    category_id_from_name(cat)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn movies_categories_match() {
        let cats = categories_for_type(LibraryType::Movies);
        assert!(cats.contains(&MOVIES_HD));
        assert_eq!(cats.len(), 9);
    }

    #[test]
    fn category_id_lookup_is_case_insensitive() {
        assert_eq!(category_id_from_name("Movies/HD"), Some(MOVIES_HD));
        assert_eq!(category_id_from_name("MOVIES/HD"), Some(MOVIES_HD));
        assert_eq!(category_id_from_name("TV/Anime"), Some(TV_ANIME));
    }

    #[test]
    fn type_for_category_buckets_by_range() {
        assert_eq!(type_for_category(2040), LibraryType::Movies);
        assert_eq!(type_for_category(5040), LibraryType::Series);
        assert_eq!(type_for_category(3000), LibraryType::Music);
        assert_eq!(type_for_category(9999), LibraryType::Other);
    }

    #[test]
    fn site_mapping_routes_to_torznab() {
        let mappings = vec![serde_json::json!({ "id": 42, "cat": "Movies/HD" })];
        assert_eq!(
            map_site_category_to_torznab(Some("42"), &mappings),
            Some(MOVIES_HD)
        );
        assert_eq!(map_site_category_to_torznab(Some("999"), &mappings), None);
    }
}

File: lib/mydia/indexers/category_mapping.ex
defmodule Mydia.Indexers.CategoryMapping do
  @moduledoc """
  Maps library types to Torznab standard categories for torrent searches.

  Torznab categories follow Newznab standard numbering:
  - 1000-1999: Console (games)
  - 2000-2999: Movies
  - 3000-3999: Audio/Music
  - 4000-4999: PC (software/games)
  - 5000-5999: TV
  - 6000-6999: XXX/Adult
  - 7000-7999: Books/Ebooks
  - 8000-8999: Other

  Both Prowlarr and Jackett use this standard category numbering.
  """

  # Torznab standard category IDs
  # Movies (2000-2999)
  @movies_parent 2000
  @movies_foreign 2010
  @movies_other 2020
  @movies_sd 2030
  @movies_hd 2040
  @movies_uhd 2045
  @movies_bluray 2050
  @movies_3d 2060
  @movies_dvd 2070

  # Audio/Music (3000-3999)
  @audio_parent 3000
  @audio_mp3 3010
  @audio_video 3020
  @audio_audiobook 3030
  @audio_lossless 3040
  @audio_other 3050
  @audio_foreign 3060

  # TV (5000-5999)
  @tv_parent 5000
  @tv_webdl 5010
  @tv_foreign 5020
  @tv_sd 5030
  @tv_hd 5040
  @tv_uhd 5045
  @tv_other 5050
  @tv_sport 5060
  @tv_anime 5070
  @tv_documentary 5080

  # Adult/XXX (6000-6999)
  @xxx_parent 6000
  @xxx_dvd 6010
  @xxx_wmv 6020
  @xxx_xvid 6030
  @xxx_x264 6040
  @xxx_uhd 6045
  @xxx_pack 6050
  @xxx_imageset 6060
  @xxx_packs 6070
  @xxx_sd 6080
  @xxx_webdl 6090

  # Books (7000-7999)
  @books_parent 7000
  @books_foreign 7010
  @books_ebook 7020
  @books_comics 7030
  @books_magazines 7040
  @books_technical 7050
  @books_other 7060
  @books_audiobook 7070

  # Other (8000-8999)
  @other_parent 8000
  @other_misc 8010
  @other_hashed 8020

  @doc """
  Returns all category IDs for the given library type.

  ## Parameters
    - library_type: One of :movies, :series, :mixed

  ## Returns
    - List of Torznab category IDs

  ## Examples

      iex> CategoryMapping.categories_for_type(:movies)
      [2000, 2010, 2020, 2030, 2040, 2045, 2050, 2060, 2070]

      iex> CategoryMapping.categories_for_type(:series)
      [5000, 5010, 5020, 5030, 5040, 5045, 5050, 5060, 5070, 5080]
  """
  @spec categories_for_type(atom()) :: [integer()]
  def categories_for_type(:movies) do
    [
      @movies_parent,
      @movies_foreign,
      @movies_other,
      @movies_sd,
      @movies_hd,
      @movies_uhd,
      @movies_bluray,
      @movies_3d,
      @movies_dvd
    ]
  end

  def categories_for_type(:series) do
    [
      @tv_parent,
      @tv_webdl,
      @tv_foreign,
      @tv_sd,
      @tv_hd,
      @tv_uhd,
      @tv_other,
      @tv_sport,
      @tv_anime,
      @tv_documentary
    ]
  end

  # Mixed library type searches across all video content
  def categories_for_type(:mixed) do
    categories_for_type(:movies) ++ categories_for_type(:series)
  end

  # Unknown types return empty list (no category filter)
  def categories_for_type(_), do: []

  @doc """
  Returns the parent category ID for a library type.

  Useful when you want to search just the main category without subcategories.

  ## Examples

      iex> CategoryMapping.parent_category(:movies)
      2000

      iex> CategoryMapping.parent_category(:series)
      5000
  """
  @spec parent_category(atom()) :: integer() | nil
  def parent_category(:movies), do: @movies_parent
  def parent_category(:series), do: @tv_parent
  def parent_category(:mixed), do: nil
  def parent_category(_), do: nil

  @doc """
  Returns category name for a given category ID.

  ## Examples

      iex> CategoryMapping.category_name(2000)
      "Movies"

      iex> CategoryMapping.category_name(3040)
      "Audio/Lossless"
  """
  @spec category_name(integer()) :: String.t()
  def category_name(id) when is_integer(id) do
    case id do
      # Movies
      @movies_parent -> "Movies"
      @movies_foreign -> "Movies/Foreign"
      @movies_other -> "Movies/Other"
      @movies_sd -> "Movies/SD"
      @movies_hd -> "Movies/HD"
      @movies_uhd -> "Movies/UHD"
      @movies_bluray -> "Movies/BluRay"
      @movies_3d -> "Movies/3D"
      @movies_dvd -> "Movies/DVD"
      # Audio
      @audio_parent -> "Audio"
      @audio_mp3 -> "Audio/MP3"
      @audio_video -> "Audio/Video"
      @audio_audiobook -> "Audio/Audiobook"
      @audio_lossless -> "Audio/Lossless"
      @audio_other -> "Audio/Other"
      @audio_foreign -> "Audio/Foreign"
      # TV
      @tv_parent -> "TV"
      @tv_webdl -> "TV/WEB-DL"
      @tv_foreign -> "TV/Foreign"
      @tv_sd -> "TV/SD"
      @tv_hd -> "TV/HD"
      @tv_uhd -> "TV/UHD"
      @tv_other -> "TV/Other"
      @tv_sport -> "TV/Sport"
      @tv_anime -> "TV/Anime"
      @tv_documentary -> "TV/Documentary"
      # Adult
      @xxx_parent -> "XXX"
      @xxx_dvd -> "XXX/DVD"
      @xxx_wmv -> "XXX/WMV"
      @xxx_xvid -> "XXX/XviD"
      @xxx_x264 -> "XXX/x264"
      @xxx_uhd -> "XXX/UHD"
      @xxx_pack -> "XXX/Pack"
      @xxx_imageset -> "XXX/ImageSet"
      @xxx_packs -> "XXX/Packs"
      @xxx_sd -> "XXX/SD"
      @xxx_webdl -> "XXX/WEB-DL"
      # Books
      @books_parent -> "Books"
      @books_foreign -> "Books/Foreign"
      @books_ebook -> "Books/EBook"
      @books_comics -> "Books/Comics"
      @books_magazines -> "Books/Magazines"
      @books_technical -> "Books/Technical"
      @books_other -> "Books/Other"
      @books_audiobook -> "Books/Audiobook"
      # Other
      @other_parent -> "Other"
      @other_misc -> "Other/Misc"
      @other_hashed -> "Other/Hashed"
      # Unknown
      _ -> "Unknown (#{id})"
    end
  end

  @doc """
  Checks if a category ID belongs to a given library type.

  ## Examples

      iex> CategoryMapping.category_matches_type?(2040, :movies)
      true

      iex> CategoryMapping.category_matches_type?(5000, :movies)
      false
  """
  @spec category_matches_type?(integer(), atom()) :: boolean()
  def category_matches_type?(category_id, library_type) do
    category_id in categories_for_type(library_type)
  end

  @doc """
  Returns the library type for a given category ID.

  ## Examples

      iex> CategoryMapping.type_for_category(2040)
      :movies

      iex> CategoryMapping.type_for_category(3000)
      :other
  """
  @spec type_for_category(integer()) :: atom()
  def type_for_category(category_id) when is_integer(category_id) do
    cond do
      category_id >= 2000 and category_id < 3000 -> :movies
      category_id >= 5000 and category_id < 6000 -> :series
      true -> :other
    end
  end

  @doc """
  Returns the Torznab category ID for a category name.

  This is used to convert Cardigann category names (like "Movies/HD" or "TV/Anime")
  to their standard Torznab numeric IDs.

  ## Examples

      iex> CategoryMapping.category_id_from_name("Movies/HD")
      2040

      iex> CategoryMapping.category_id_from_name("TV/Anime")
      5070

      iex> CategoryMapping.category_id_from_name("Unknown")
      nil
  """
  @spec category_id_from_name(String.t()) :: integer() | nil
  def category_id_from_name(name) when is_binary(name) do
    # Normalize the name for lookup (case-insensitive)
    normalized = String.downcase(name)

    case normalized do
      # Movies
      "movies" -> @movies_parent
      "movies/foreign" -> @movies_foreign
      "movies/other" -> @movies_other
      "movies/sd" -> @movies_sd
      "movies/hd" -> @movies_hd
      "movies/uhd" -> @movies_uhd
      "movies/bluray" -> @movies_bluray
      "movies/blu-ray" -> @movies_bluray
      "movies/3d" -> @movies_3d
      "movies/dvd" -> @movies_dvd
      "movies/divx/xvid" -> @movies_sd
      "movies/h.264/x264" -> @movies_hd
      "movies/hevc/x265" -> @movies_hd
      "movies/mp4" -> @movies_sd
      "movies/dubs/dual audio" -> @movies_foreign
      "movies/bollywood" -> @movies_foreign
      "movies/svcd/vcd" -> @movies_sd
      # Audio
      "audio" -> @audio_parent
      "audio/mp3" -> @audio_mp3
      "audio/video" -> @audio_video
      "audio/audiobook" -> @audio_audiobook
      "audio/lossless" -> @audio_lossless
      "audio/other" -> @audio_other
      "audio/foreign" -> @audio_foreign
      "music" -> @audio_parent
      "music/mp3" -> @audio_mp3
      "music/lossless" -> @audio_lossless
      "music/video" -> @audio_video
      "music/dvd" -> @audio_parent
      "music/radio" -> @audio_other
      "music/other" -> @audio_other
      "music/album" -> @audio_parent
      "music/box set" -> @audio_parent
      "music/discography" -> @audio_parent
      "music/single" -> @audio_parent
      "music/concerts" -> @audio_video
      "music/aac" -> @audio_mp3
      # TV
      "tv" -> @tv_parent
      "tv/web-dl" -> @tv_webdl
      "tv/webdl" -> @tv_webdl
      "tv/foreign" -> @tv_foreign
      "tv/sd" -> @tv_sd
      "tv/hd" -> @tv_hd
      "tv/uhd" -> @tv_uhd
      "tv/other" -> @tv_other
      "tv/sport" -> @tv_sport
      "tv/anime" -> @tv_anime
      "tv/documentary" -> @tv_documentary
      "tv/dvd" -> @tv_parent
      "tv/divx/xvid" -> @tv_sd
      "tv/svcd/vcd" -> @tv_sd
      "tv/hevc/x265" -> @tv_hd
      "tv/cartoons" -> @tv_anime
      # Adult
      "xxx" -> @xxx_parent
      "xxx/dvd" -> @xxx_dvd
      "xxx/wmv" -> @xxx_wmv
      "xxx/xvid" -> @xxx_xvid
      "xxx/x264" -> @xxx_x264
      "xxx/uhd" -> @xxx_uhd
      "xxx/pack" -> @xxx_pack
      "xxx/imageset" -> @xxx_imageset
      "xxx/packs" -> @xxx_packs
      "xxx/sd" -> @xxx_sd
      "xxx/web-dl" -> @xxx_webdl
      "xxx/video" -> @xxx_dvd
      "xxx/picture" -> @xxx_imageset
      "xxx/magazine" -> @xxx_parent
      "xxx/hentai" -> @xxx_parent
      "xxx/games" -> @xxx_parent
      # Books
      "books" -> @books_parent
      "books/foreign" -> @books_foreign
      "books/ebook" -> @books_ebook
      "books/comics" -> @books_comics
      "books/magazines" -> @books_magazines
      "books/technical" -> @books_technical
      "books/other" -> @books_other
      "books/audiobook" -> @books_audiobook
      # Other
      "other" -> @other_parent
      "other/misc" -> @other_misc
      "other/hashed" -> @other_hashed
      "other/emulation" -> @other_parent
      "other/tutorial" -> @books_parent
      "other/sounds" -> @audio_other
      "other/e-books" -> @books_ebook
      "other/images" -> @other_parent
      "other/mobile phone" -> @other_parent
      "other/comics" -> @books_comics
      "other/other" -> @other_misc
      "other/nulled script" -> @other_parent
      "other/audiobook" -> @audio_audiobook
      # PC/Console categories (mapped to Other for now)
      "pc" -> @other_parent
      "pc/games" -> @other_parent
      "pc/mac" -> @other_parent
      "pc/mobile-android" -> @other_parent
      "pc/mobile-ios" -> @other_parent
      "pc/mobile-other" -> @other_parent
      "console" -> @other_parent
      "console/ps3" -> @other_parent
      "console/ps4" -> @other_parent
      "console/psp" -> @other_parent
      "console/xbox" -> @other_parent
      "console/xbox 360" -> @other_parent
      "console/wii" -> @other_parent
      "console/nds" -> @other_parent
      "console/3ds" -> @other_parent
      "console/other" -> @other_parent
      # Not found
      _ -> nil
    end
  end

  def category_id_from_name(_), do: nil

  @doc """
  Maps a site-specific category ID to a Torznab category ID using the provided category mappings.

  ## Parameters
    - site_category_id: The site-specific category ID (as string or integer)
    - category_mappings: List of category mapping entries from the Cardigann definition

  ## Returns
    - The Torznab category ID, or nil if no mapping found

  ## Examples

      iex> mappings = [%{"id" => 42, "cat" => "Movies/HD"}]
      iex> CategoryMapping.map_site_category_to_torznab("42", mappings)
      2040

      iex> CategoryMapping.map_site_category_to_torznab("999", mappings)
      nil
  """
  @spec map_site_category_to_torznab(String.t() | integer() | nil, list()) :: integer() | nil
  def map_site_category_to_torznab(nil, _mappings), do: nil
  def map_site_category_to_torznab(_site_id, nil), do: nil
  def map_site_category_to_torznab(_site_id, []), do: nil

  def map_site_category_to_torznab(site_category_id, category_mappings) do
    # Normalize site_category_id to string for comparison
    site_id_str = to_string(site_category_id)
    site_id_int = parse_category_id(site_category_id)

    # Find matching category mapping
    mapping =
      Enum.find(category_mappings, fn mapping ->
        mapping_id = Map.get(mapping, "id") || Map.get(mapping, :id)
        to_string(mapping_id) == site_id_str || mapping_id == site_id_int
      end)

    case mapping do
      nil ->
        nil

      mapping ->
        # Get the Torznab category name and convert to ID
        cat_name = Map.get(mapping, "cat") || Map.get(mapping, :cat)
        category_id_from_name(cat_name)
    end
  end

  defp parse_category_id(id) when is_integer(id), do: id

  defp parse_category_id(id) when is_binary(id) do
    case Integer.parse(id) do
      {num, _} -> num
      :error -> nil
    end
  end

  defp parse_category_id(_), do: nil

  @doc """
  Returns all defined Torznab categories with their IDs and names.

  Useful for UI displays or debugging.
  """
  @spec all_categories() :: [%{id: integer(), name: String.t(), type: atom()}]
  def all_categories do
    all_ids()
    |> Enum.map(fn id ->
      %{
        id: id,
        name: category_name(id),
        type: type_for_category(id)
      }
    end)
  end

  # Private: All defined category IDs
  defp all_ids do
    [
      # Movies
      @movies_parent,
      @movies_foreign,
      @movies_other,
      @movies_sd,
      @movies_hd,
      @movies_uhd,
      @movies_bluray,
      @movies_3d,
      @movies_dvd,
      # Audio
      @audio_parent,
      @audio_mp3,
      @audio_video,
      @audio_audiobook,
      @audio_lossless,
      @audio_other,
      @audio_foreign,
      # TV
      @tv_parent,
      @tv_webdl,
      @tv_foreign,
      @tv_sd,
      @tv_hd,
      @tv_uhd,
      @tv_other,
      @tv_sport,
      @tv_anime,
      @tv_documentary,
      # Adult
      @xxx_parent,
      @xxx_dvd,
      @xxx_wmv,
      @xxx_xvid,
      @xxx_x264,
      @xxx_uhd,
      @xxx_pack,
      @xxx_imageset,
      @xxx_packs,
      @xxx_sd,
      @xxx_webdl,
      # Books
      @books_parent,
      @books_foreign,
      @books_ebook,
      @books_comics,
      @books_magazines,
      @books_technical,
      @books_other,
      @books_audiobook,
      # Other
      @other_parent,
      @other_misc,
      @other_hashed
    ]
  end
end
